### PR TITLE
feat: add AI translation support 

### PR DESCRIPTION
**Pull request by loorr**
## Summary

  Add AI-powered Chinese translation support to Markdown Preview Enhanced. #2344

  ## Changes

  - Add commands to translate the current preview and restore the original content.
  - Support streaming and incremental block-level translation.
  - Cache translations and only retranslate changed blocks.
  - Store API keys securely using VS Code SecretStorage.
  - Add configurable AI provider, model, and automatic update settings.
  - Update crossnote to the packaged `0.9.32` release.
  - Bump the extension version to `0.8.31`.

  ## Related PR

  - crossnote translation context-menu support: https://github.com/shd101wyy/crossnote/pull/459

  ## Dependency note

  This PR temporarily uses the crossnote `0.9.32` release artifact from the fork. Once the related upstream changes are released, the dependency can be changed to:

### CODE REVIEW

File: src/ai-translation-cache.ts [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+ * Inputs: the source markdown, the provider id, the model id, and a
+ * system-prompt version tag (so changing the prompt invalidates the cache).
+ */
+export function computeTranslationCacheKey(
+  markdown: string,
+  provider: string,
+  model: string,
+): string {
+  const hash = crypto
+    .createHash('sha256')
```
Comment (shd101wyy) [inline src/ai-translation-cache.ts]: Blocking: this breaks the web extension build. In the browser bundle, `esbuild-plugin-polyfill-node` maps `node:crypto` to an empty module — the web build warns `Import "createHash" will always be undefined` here (and in markdown-blocks.ts), so at runtime on vscode.dev `computeTranslationCacheKey` throws and the translate command dies as a silent unhandled rejection. `crypto.subtle.digest` (WebCrypto) works in Node 18+ and browsers alike; alternatively gate the commands with `"enablement": "!isWeb"`. Verified by building both targets in a clean checkout.

File: src/ai-translation-cache.ts [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+  /** The translated markdown (fed back into crossnote's preview pipeline). */
+  markdown: string;
+  provider: string;
+  model: string;
+  createdAt: number;
+  sourceBytes: number;
+}
+
+/** Read a cached translation. Returns undefined on miss / read error. */
+export function getCachedTranslation(
```
Comment (shd101wyy) [inline src/ai-translation-cache.ts]: KISS (recommended, non-blocking): the whole-document cache is redundant given the block cache — `translateIncrementally` already assembles the document with zero API calls when every block is cached, which covers the unchanged-document fast path. Dropping the whole-doc cache removes ~100 lines, the `.meta.json` duplication, and the only unbounded disk-growth path (the blocks dir has a 5000-entry LRU; these `.md` files accumulate forever with no eviction or clear command).

File: src/ai-translator.ts [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+  ): AsyncIterable<{ type: string; delta?: string }>;
+};
+
+let modelsPromise: Promise<PiModels> | undefined;
+
+async function getModels(): Promise<PiModels> {
+  if (!modelsPromise) {
+    modelsPromise = (async () => {
+      const { builtinModels } =
+        await import('@earendil-works/pi-ai/providers/all');
```
Comment (shd101wyy) [inline src/ai-translator.ts:81]: Blocking: `tsc --noEmit` fails here — `TS2307: Cannot find module '@earendil-works/pi-ai/providers/all'` under the repo's current `moduleResolution` (the package is ESM-only with an exports map). Runtime is fine because esbuild follows the exports map, but the typecheck needs a fix: re-export the needed symbols from the package root, add a local `.d.ts` for the subpath, or move the repo to `moduleResolution: "bundler"`.

File: src/config.ts [hunk 4]
```diff
@@ -45,7 +45,9 @@ type VSCodeMPEConfigKey =
   | 'd2Path'
   | 'd2Layout'
   | 'd2Theme'
-  | 'd2Sketch';
+  | 'd2Sketch'
+  | 'aiTranslationProvider'
+  | 'aiTranslationModel';
```
Comment (shd101wyy) [inline src/config.ts]: Blocking: `aiTranslationAutoUpdate` is missing from this union, so `getMPEConfig<boolean>('aiTranslationAutoUpdate')` in preview-provider.ts fails `tsc --noEmit` (TS2345). Please add `| 'aiTranslationAutoUpdate'`.

File: src/preview-provider.ts [hunk 5]
```diff
@@ -907,6 +978,587 @@ export class PreviewProvider {
...
+    } catch {
+      return;
+    }
+    if (!markdown.trim()) {
+      return;
+    }
+
+    const providerId = getMPEConfig<string>('aiTranslationProvider') ?? '';
+    const modelId = getMPEConfig<string>('aiTranslationModel') ?? '';
+    if (!providerId || !modelId) {
```
Comment (shd101wyy) [inline src/preview-provider.ts:1050]: Blocking (UX): this silently no-ops when provider/model aren't configured, and every downstream failure is silent too — the helpful error strings in `resolveTranslationRequest` and `result.error` are only ever written to the Output channel. From the user's perspective, the command does nothing. Please show an error message (with a hint toward the Set API Key command / settings) on the user-visible paths.

File: src/preview-provider.ts [hunk 6]
```diff
@@ -907,6 +978,587 @@ export class PreviewProvider {
...
+      },
+    );
+
+    if (this.abortControllers.get(sourceUri.toString()) === controller) {
+      this.abortControllers.delete(sourceUri.toString());
+    }
+    if (controller.signal.aborted) {
+      return undefined;
+    }
+    if (!result.ok || !result.markdown) {
```
Comment (shd101wyy) [inline src/preview-provider.ts]: Same theme: `result.error` is dropped here (and in `translateWholeFallback`). At minimum log it — right now a failing 'Translating…' notification just closes with no explanation.

File: src/preview-provider.ts [hunk 7]
```diff
@@ -907,6 +978,587 @@ export class PreviewProvider {
...
+    return result.markdown;
+  }
+
+  /**
+   * Whole-document fallback: a single streaming translation of the entire
+   * markdown. Used when block splitting fails or every per-block translation
+   * fails. Returns the translated markdown, or `undefined` if aborted or
+   * failed.
+   */
+  private async translateWholeFallback(
```
Comment (shd101wyy) [inline src/preview-provider.ts]: DRY (non-blocking): this is a near-exact duplicate of `translateWholeDocumentStreaming` — the only difference is `onPartial`. One `streamWholeDocument({ markdown, onPartial? })` helper would remove ~40 lines and a second AbortController/withProgress dance to keep in sync.

File: src/preview-provider.ts [hunk 8]
```diff
@@ -608,10 +661,8 @@ export class PreviewProvider {
       ) {
         return;
       }
-
       previewPanel.webview.html = html;
     } catch (error) {
-      vscode.window.showErrorMessage(String(error));
```
Comment (shd101wyy) [inline src/preview-provider.ts:659]: Unrelated change — please revert. Swallowing render errors here makes preview regressions much harder to debug.

**Review by shd101wyy — CHANGES_REQUESTED** [review]
Thanks for the well-structured PR, and for correctly splitting the webview-side changes into shd101wyy/crossnote#459 — the architecture is right (AI calls, secrets, and settings live in the extension host; the translated markdown is fed back through crossnote's normal `initPreview` pipeline so TOC/exports/line-numbers keep working; the context-menu hooks live in crossnote).

I checked out this branch in a clean worktree and built + typechecked it. The security design is solid (details below), but there are two verified build/typecheck breakages and a fork-dependency issue that block merging, plus a KISS/DRY pass worth doing:

## Blocking

1. **Web extension build is broken.** In the browser bundle, `esbuild-plugin-polyfill-node` maps `node:crypto` to an *empty* module. The web build emits `[WARNING] Import "createHash" will always be undefined` for `ai-translation-cache.ts:34` and `markdown-blocks.ts:160`, so on vscode.dev the translate command throws (`createHash is not a function`) and dies as a **silent unhandled rejection**. Fix with WebCrypto (`crypto.subtle.digest` works in Node 18+ and browsers) or gate the feature with `"enablement": "!isWeb"`.
2. **`npx tsc --noEmit` fails with 2 errors** — `aiTranslationAutoUpdate` is missing from the `VSCodeMPEConfigKey` union in `src/config.ts`, and `@earendil-works/pi-ai/providers/all` doesn't resolve under the repo's current `moduleResolution`. (Heads-up: `check:all` only runs eslint+prettier, so CI won't catch this — strict tsc is still a repo convention.)
3. **The dependency on the fork tarball `loorr/crossnote@0.9.32` blocks merging.** The bytes are hash-pinned in yarn.lock, and I verified the installed fork contains essentially just upstream + crossnote#459 — so this is process, not trust. But we can't ship a forked renderer indefinitely. Please rebase this onto the official npm package once shd101wyy/crossnote#459 is merged and released (npm is at 0.9.31 today).
4. **All failure paths are silent.** Missing config, missing API key, and API/stream errors are dropped — the user clicks Translate and nothing happens. Please surface `result.error` / the `resolveTranslationRequest` messages via `showErrorMessage`, and revert the unrelated removal of `showErrorMessage` in the render-error catch.

## Security review (for the record — no action needed)

- API key: SecretStorage ✓, password-masked input ✓, used only in the extension host ✓, never posted to the webview ✓, never logged (Output channel logs sizes/counts only) ✓.
- Translated markdown renders through the standard crossnote pipeline, so no new XSS surface vs. what the preview already does with a document.
- `aiTranslationAutoUpdate` defaulting to `false` avoids surprise API spend 👍
- `@earendil-works/pi-ai` is a reputable package (mitsuhiko/badlogic), loaded lazily, invoked only on explicit user action.

## Non-blocking (KISS / DRY)

- The whole-document cache is redundant given the block cache (inline). Dropping it deletes ~100 lines and removes the only unbounded disk-growth path.
- `translateWholeDocumentStreaming` vs `translateWholeFallback` are near-duplicates (inline).
- If you want to cut the most complex part: the streaming partial-render machinery (`refreshChains` serialization + run-by-run `updateHtml` refreshes) could be replaced by "progress notification, render once at the end" at the cost of the live-streaming UX. Your call.
- `clearAiApiKey` is dead code — there's no command to clear the key or the caches. Consider a combined "clear" command.
- With `aiTranslationAutoUpdate` off, edits made while in translation mode are invisible until Show Original — worth stating in the setting description.
- `splitMarkdownBlocks`' fence handling is edge-case-prone; the repo has a mocha unit-test pattern (`test/*.test.js`) — tests here would be genuinely valuable.
- Process: please move the version bump and dated CHANGELOG section under `[Unreleased]` — releases here are cut by the maintainer.

Happy to re-review quickly once the blocking items are in. Thanks again!
